### PR TITLE
Added the ability to filter the list of locales in the Remap tab.

### PR DESCRIPTION
**Pull request by DmitriySalnikov**
With this PR no longer need to scroll through the list of all ~350 locales supported by godot.

<details>
<summary>With filter</summary>
<a href="https://user-images.githubusercontent.com/7782218/31676314-e3887846-b36f-11e7-8f69-54893a53b606.png" target="_blank"></a>
</br>
<a href="https://user-images.githubusercontent.com/7782218/31676323-ec810d1e-b36f-11e7-9990-04eea4386b0e.png" target="_blank"></a>
</details>
<details>
<summary>Without filter</summary>
<a href="https://user-images.githubusercontent.com/7782218/31676330-f1aca50a-b36f-11e7-9f78-b6df508f1f41.png" target="_blank"></a>
</details>

### CODE REVIEW

**Comment by groud** [comment]
Why not putting that in the Localization tab ?

**Comment by DmitriySalnikov** [comment]
@groud good idea. thank you.
I'll go fix it.

**Comment by DmitriySalnikov** [comment]
@groud where would you like to see this setting?
godot64_2017-10-17_22-08-28

**Comment by groud** [comment]
Well in fact I may have a better solution :)

Instead of adding a new option in the editor filtering the languages, we could display the already used languages (for other resources) at the top of the popup. We could also reduce the height of the popup.

What do you think ?

**Comment by DmitriySalnikov** [comment]
@groud I'm not sure it will be comfortable.

Also, the project will support only the languages for which the developer or team will be able to translate their game.
They are not randomly selected from the list of languages. It will be a thoughtful list of supported languages.

I'm thinking about adding a button with a pop-up window where you can easily choose the desired languages. Something like this(but with different buttons):
godot windows opt tools 64_2017-10-17_23-22-17
Or leave as is.

**Comment by groud** [comment]
Yeah, sure going up the popup would not be comfortable. But if you reduce the height of the popup it's ok.

The point is that adding this option is bloating the configuration menu for little gain. My philisophy on that is avoiding options that impact other options. Usually, when you have to do that it means that your GUI is not well designed.

An other approch would be to keep this filter, but remove the selection dropdown: if you select "fr, en, es" you've got in the bottom view a list with only those three locales with either a file name (if the remap is set) or nothing (if not). With an edit button to change the value.

**Comment by DmitriySalnikov** [comment]
@groud 
Hmm
What do you think if I rename this parameter and make it available in the game?
For Example: "Supported Languages"
And add function to retrieve list of supported languages.
And maybe add functions to modify this list directly from the game, for example, to support mods.
Like this:
get_supported_locales()
get_supported_locale_names()
set_supported_locales(Array)
is_supported_locale(String)
remove_supported_locale(String)
append_supported_locale(String)

> An other approch would be to keep this filter, but remove the selection dropdown: if you select "fr, en, es" you've got in the bottom view a list with only those three locales with either a file name (if the remap is set) or nothing (if not). With an edit button to change the value.

Yes, i know. This is because the filtered list does not contain titles that have been removed.

**Comment by DmitriySalnikov** [comment]
>But if you reduce the height of the popup it's ok.

I think it's impossible.

Need help and advice how best to do this PR

**Comment by groud** [comment]
> I think it's impossible.

It probably can be modified into tree.cpp (line 1779) I guess.

**Comment by DmitriySalnikov** [comment]
> It probably can be modified into tree.cpp (line 1779) I guess.

Nope. It doesn't work. And that's a completely different class that requires a separate PR. So you can not just change the size of the pop-up window.
What about this [comment](https://github.com/godotengine/godot/pull/12186#issuecomment-337409931)?

**Comment by DmitriySalnikov** [comment]
@akien-mga maybe you have an idea of how best to do?

**Comment by DmitriySalnikov** [comment]
@akien-mga maybe you have an idea of how best to do?

File: editor/project_settings_editor.cpp [hunk 1]
```diff
@@ -1127,7 +1127,12 @@ void ProjectSettingsEditor::_translation_res_option_changed() {
 	ERR_FAIL_COND(!remaps.has(key));
 	PoolStringArray r = remaps[key];
 	ERR_FAIL_INDEX(idx, r.size());
-	r.set(idx, path + ":" + langs[which]);
+	if(filtered_locales_idxs_remap.size()>0){
+		r.set(idx, path + ":" + langs[filtered_locales_idxs_remap[which]]);
+    }
+    else{
+        r.set(idx, path + ":" + langs[which]);
+    }
```
Comment (akien-mga) [inline editor/project_settings_editor.cpp]: The style is completely wrong here. See http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

**Comment by reduz** [comment]
I suggest something like this, in the locales tab:

image

**Comment by NathanWarden** [comment]
I disagree. Jedi speak English, Spanish, etc last I checked, so we should just use those languages, but Klingon and Ogre make sense to keep in the list.

**Comment by DmitriySalnikov** [comment]
image
working functionality

**Comment by DmitriySalnikov** [comment]
check it out please

**Comment by akien-mga** [comment]
Could you squash the commits together? See http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

**Comment by DmitriySalnikov** [comment]
fixed, I hope

**Pull request merged**